File: problems/10-graphs/002-number-of-provinces/java/Solution.java
public class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        boolean[] visited = new boolean[n];
        int provinces = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                provinces++;
                dfs(i, isConnected, visited);
            }
        }

        return provinces;
    }

    private void dfs(int node, int[][] graph, boolean[] visited) {
        visited[node] = true;

        for (int next = 0; next < graph.length; next++) {
            if (graph[node][next] == 1 && !visited[next]) {
                dfs(next, graph, visited);
            }
        }
    }
}

File: problems/10-graphs/002-number-of-provinces/python/solution.py
class Solution:
    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        n = len(isConnected)
        visited = [False] * n
        provinces = 0

        def dfs(node: int) -> None:
            visited[node] = True

            for nxt in range(n):
                if isConnected[node][nxt] == 1 and not visited[nxt]:
                    dfs(nxt)

        for i in range(n):
            if not visited[i]:
                provinces += 1
                dfs(i)

        return provinces
